crates/noirc_evaluator/src/ssa_refactor/opt/flatten_cfg.rs

```diff
@@ -0,0 +1,566 @@
+//! The flatten cfg optimization pass "flattens" the entire control flow graph into a single block.
+//! This includes branches in the CFG with non-constant conditions. Flattening these requires
+//! special handling for operations with side-effects and can lead to a loss of information since
+//! the jmpif will no longer be in the program. As a result, this pass should usually be towards or
+//! at the end of the optimization passes. Note that this pass will also perform unexpectedly if
+//! loops are still present in the program. Since the pass sees a normal jmpif, it will attempt to
+//! merge both blocks, but no actual looping will occur.
+//!
+//! This pass is also known to produce some extra instructions which may go unused (usually 'Not')
+//! while merging branches. These extra instructions can be cleaned up by a later dead instruction
+//! elimination (DIE) pass.
```

Comment:
DIE shouldn't touch store instructions since removing those requires a different algorithm than removing other instructions where you can just look to see if their results are unused.

I would expect the mem2reg pass to remove any store instructions afterward however.